### DISCUSSION TITLE
How to know whether state of nested machine is changed?

### DISCUSSION THREAD

**Discussion by gsbelarus** [statelyai/xstate#1226] [General]
I have a complex machine with several nested sub-machines invoked as following:
```ts
          ...
          enterDateBegin2: {
            invoke: {
              id: 'calendarMachine',
              src: calendarMachine,
              autoForward: true,
              data: (ctx: IBotMachineContext) => ({
                selectedDate: ctx.dateBegin2,
                canceled: false,
                ...
              }),
              onDone: [
                {
                  cond: (_, event) => event.data.canceled,
                  target: '#botMachine.mainMenu'
                },
                {
                  target: 'showComparePayslip',
                  actions: assign({ dateBegin2: (_, event) => event.data.selectedDate })
                }
              ]
            }
          },
         ...
```
The machine is run with *interpreter*.

I need to know after every event sent to machine whether its state has been changed. While we stay inside main machine the property *state.changed* of interpreter works well. But when events are passed to invoked nested machine state of the parent machine remains unchanged i.e. `state.changed === false`.

How can I know from given situation that the state of nested machine is changed as a result of passing event? What property should I test? I tried *children* of interpreter but it remains empty.

**Comment by gsbelarus** [comment]
As for now I have found the following solution:
```ts
      service.send(e);

      let childrenStateChanged = false;

      for (const [_key, ch] of service.children) {
        if (ch.state.changed) {
          childrenStateChanged = true;
          break;
        }
      }

      if (!service.state.changed && !childrenStateChanged) {
```
It looks quite cumbersome. Do we need some easier way of finding did sended event changed state of the machine or some of its nested machines or not.

**Comment by Andarist** [comment]
A more interesting question is - why do you need this? It would be best to always describe a concrete use case when asking this kind of question. It lets others get a better understanding of your problem.

**Comment by davidkpiano** [comment]
You aren't supposed to know that directly from the parent machine. If it seems difficult, there's a reason it's difficult (and not encouraged).

Remember: the main communication mechanism is **through events**. If you want to know something, you do it through an event.

Imagine you made an API request that caused some remote server to do a bunch of things before returning a response. How is the client supposed to know when the remote server "changes its state"? The answer is that it _shouldn't_ know, unless it is explicitly told that it changed.

What is your use-case for knowing if a child service changed?

**Comment by gsbelarus** [comment]
I use state machine for programming a Telegram/Viber chat-bot. User interacts with the bot through menu commands or entering some text. I send user input to the machine using actions like `{ type: 'ENTER_TEXT', text: ... }` or `{ type: 'MENU_COMMAND', command: ... }. Inside the machine the decision is made via guards and context is updated via actions based on the data (text, command) provided. 

As user-bot dialogue resides inside the messenger program it could easily be disconnected from the machine's internal state. For example, user was in the middle of multi step dialogue and left the chat for a few days. During this period of time the server could be restarted and internal state of the machine could be lost. So, when the user returns and tries to continue the process, events, sent to machine, would not change its state as this state is not suitable for them anymore. Then, I want to detect such situation and ask user to restart dialogue.

Of course, the better solution would be to save/restore machine state. But, as I use nested submachines the automatic restoration is impossible and manual seems to complex for me.

Also, consider that there would be up to 10 000 simultaneous active chats. As persisted state object is so comprehensive (actions, events, history etc) it could be a problem to serialize and save so many JSON-s.

**Comment by Andarist** [comment]
So it seems that for each command you expect some answer - based on which you want to display a dialog or not. The best way to handle this would be to provide this answer explicitly to the requester (parent machine in this case) in the form of an event. Note how this would nicely embed information about this dialog in the parent's machine (which could be visualized) whereas when you depend on .changed this information is lost for the visualization.

**Reply by gsbelarus** [reply]
Not exactly, my problem is that user's response (entered text or pressed menu button) could be not relevant to the current state of the machine any more. I want to detect such "wrong" input and send to user information, something like: "Sorry, we are lost a little bit. Could you,  please, start it over."

The only other solution I see is to process ENTER_TEXT and MENU_COMMAND events at **EVERY** machine's state. But then the code will become too bulky.

**Reply by Andarist** [reply]
It's, of course, hard to give concrete advice without knowing more about your implementation but the first thing that comes to my mind is that you could split your child machine into several top-level states - in one you would know that you want to handle those 2 events and in a second one you would know that you would always want to "ignore them" (as in - respond back with "we are lost"). I don't know what other states etc you have there but I imagine this being possible:

```js
const states = {
    /* states/events that have to be handled at all times */
}
const machine = createMachine({
    states: {
        accepting_commands: {
            /* spread states and add your ENTER_TEXT and MENU_COMMAND handlers */
        },
        not_accepting_commands: {
            ...states,
            on: {
                ENTER_TEXT: { actions: [sendParent('ASK_FOR_RESTART')] },
                MENU_COMMAND: { actions: [sendParent('ASK_FOR_RESTART')] }
            }
        }
    }
})
```

If you could share your child machine config this would be helpful for us to understand its structure and overall problem.

**Reply by gsbelarus** [reply]
The whole machine and nested submachine are here: https://github.com/gsbelarus/gdmn-bot/blob/master/src/machine.ts

**Reply by Andarist** [reply]
So it seems that something like this should be enough:
https://github.com/Andarist/gdmn-bot/commit/1c94d87fc056ec13c9bd1ed96d796ff78821995c

Basically, if this machine receives one of those events in a state that has a transition defined for it then it will be handled there and **won't "bubble up"** to those "root" handlers, but if it receives such event in a state that has no transition defined for it then it **will "buble up"** to those "root" handlers and the parent will get notified about thanks to `sendParent`.

Keep in mind that using `autoForward` on those invoked machines might not be really what you want because any event sent from such a child to a parent will "circle back" to the sender. They don't have transitions defined for it so such event will be ignored, but nevertheless this could surprise you in the future, even if only when debugging something.
